crates/lunatic-metrics-api/src/lib.rs

```diff
@@ -1,161 +1,750 @@
+    linker.func_wrap("lunatic::metrics", "meter_drop", meter_drop)?;
+
+    linker.func_wrap("lunatic::metrics", "event", event)?;
+
     linker.func_wrap("lunatic::metrics", "counter", counter)?;
-    linker.func_wrap("lunatic::metrics", "increment_counter", increment_counter)?;
-    linker.func_wrap("lunatic::metrics", "gauge", gauge)?;
-    linker.func_wrap("lunatic::metrics", "increment_gauge", increment_gauge)?;
-    linker.func_wrap("lunatic::metrics", "decrement_gauge", decrement_gauge)?;
+    linker.func_wrap("lunatic::metrics", "counter_add", counter_add)?;
+    linker.func_wrap("lunatic::metrics", "counter_drop", counter_drop)?;
+
+    linker.func_wrap("lunatic::metrics", "up_down_counter", up_down_counter)?;
+    linker.func_wrap(
+        "lunatic::metrics",
+        "up_down_counter_add",
+        up_down_counter_add,
+    )?;
+    linker.func_wrap(
+        "lunatic::metrics",
+        "up_down_counter_drop",
+        up_down_counter_drop,
+    )?;
+
     linker.func_wrap("lunatic::metrics", "histogram", histogram)?;
+    linker.func_wrap("lunatic::metrics", "histogram_record", histogram_record)?;
+    linker.func_wrap("lunatic::metrics", "histogram_drop", histogram_drop)?;
+
     Ok(())
 }
 
-fn get_string_arg<T>(
-    caller: &mut Caller<T>,
-    name_str_ptr: u32,
-    name_str_len: u32,
-    func_name: &str,
-) -> Result<String> {
-    let memory = get_memory(caller)?;
-    let memory_slice = memory.data(caller);
-    let name = memory_slice
-        .get(name_str_ptr as usize..(name_str_ptr + name_str_len) as usize)
-        .or_trap(func_name)?;
-    let name = String::from_utf8(name.to_vec()).or_trap(func_name)?;
-    Ok(name)
+/// Starts a new span of work, used for recording metrics including log events and meters such as counters, gauges, and histograms.
+///
+/// If parent is set to u64::MAX, then the last created span will be used.
+///
+/// Traps:
+/// * If the name is not a valid utf8 string.
+/// * If the attributes is not valid json.
+/// * If the parent span does not exist.
+/// * If any memory outside the guest heap space is referenced.
+fn span_start<T>(
+    mut caller: Caller<'_, T>,
+    parent: u64,
+    name_ptr: u32,
+    name_len: u32,
+    attributes_ptr: u32,
+    attributes_len: u32,
+) -> Result<u64>
+where
+    T: MetricsCtx,
+    <<T as MetricsCtx>::Tracer as Tracer>::Span: Send + Sync + 'static,
+{
+    let memory = get_memory(&mut caller)?;
+    let (data, state) = memory.data_and_store_mut(&mut caller);
+
+    let parent = if parent != u64::MAX {
+        Some(parent)
+    } else {
+        None
+    };
+
+    let name = get_string_arg(data, name_ptr, name_len).or_trap("lunatic::metrics::span_start")?;
+    let attributes = if attributes_len > 0 {
+        let attributes_data = data
+            .get(attributes_ptr as usize..(attributes_ptr + attributes_len) as usize)
+            .or_trap("lunatic::metrics::span_start")?;
+        let attributes_json =
```

Comment:
Should vm include some attrs like env/proc id? We should make sure guests cannot override them. This is useful in a multi-tenant/hosted case where don't trust guests.

Reply:
Yep I'll work on adding this

Reply:
Added this in https://github.com/lunatic-solutions/lunatic/pull/202/commits/4019049d31bd4d2a8c75f9875a9b1be243d64e8f